pkg/sdkclient/batchmapper/client.go

```diff
@@ -0,0 +1,180 @@
+
+import (
+	"context"
+	"errors"
+	"io"
+
+	"google.golang.org/grpc"
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	batchmappb "github.com/numaproj/numaflow-go/pkg/apis/proto/batchmap/v1"
+	"github.com/numaproj/numaflow-go/pkg/info"
+
+	"github.com/numaproj/numaflow/pkg/sdkclient"
+	sdkerr "github.com/numaproj/numaflow/pkg/sdkclient/error"
+	grpcutil "github.com/numaproj/numaflow/pkg/sdkclient/grpc"
+)
+
+// client contains the grpc connection and the grpc client.
+type client struct {
+	conn    *grpc.ClientConn
+	grpcClt batchmappb.BatchMapClient
+}
+
+// New creates a new client object.
+func New(serverInfo *info.ServerInfo, inputOptions ...sdkclient.Option) (Client, error) {
+	var opts = sdkclient.DefaultOptions(sdkclient.BatchMapAddr)
+
+	for _, inputOption := range inputOptions {
+		inputOption(opts)
+	}
+
+	// Connect to the server
+	conn, err := grpcutil.ConnectToServer(opts.UdsSockAddr(), serverInfo, opts.MaxMessageSize())
+	if err != nil {
+		return nil, err
+	}
+
+	c := new(client)
+	c.conn = conn
+	c.grpcClt = batchmappb.NewBatchMapClient(conn)
+	return c, nil
+}
+
+func NewFromClient(c batchmappb.BatchMapClient) (Client, error) {
+	return &client{
+		grpcClt: c,
+	}, nil
+}
+
+// CloseConn closes the grpc client connection.
+func (c *client) CloseConn(ctx context.Context) error {
+	return c.conn.Close()
+}
+
+// IsReady returns true if the grpc connection is ready to use.
+func (c *client) IsReady(ctx context.Context, in *emptypb.Empty) (bool, error) {
+	resp, err := c.grpcClt.IsReady(ctx, in)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetReady(), nil
+}
+
+// BatchMapFn is the handler for the gRPC client (Numa container)
+// It takes in a stream of input Requests, sends them to the gRPC server(UDF) and then streams the
+// responses received back on a channel asynchronously.
+// We spawn 2 goroutines here, one for sending the requests over the stream
+// and the other one for receiving the responses
+func (c *client) BatchMapFn(ctx context.Context, inputCh <-chan *batchmappb.BatchMapRequest) (<-chan *batchmappb.BatchMapResponse, <-chan error) {
+	// errCh is used to track and propagate any errors that might occur during the rpc lifecyle, these could include
+	// errors in sending, UDF errors etc
+	// These are propagated to the applier for further handling
+	errCh := make(chan error)
+
+	// response channel for streaming back the results received from the gRPC server
+	// TODO(map-batch): Check if keeping this buffered would help
+	responseCh := make(chan *batchmappb.BatchMapResponse)
+
+	// BatchMapFn is a bidirectional streaming RPC
+	// We get a Map_BatchMapFnClient object over which we can send the requests,
+	// receive the responses asynchronously.
+	// TODO(map-batch): this creates a new gRPC stream for every batch,
+	// it might be useful to see the performance difference between this approach
+	// and a long-running RPC
+	stream, err := c.grpcClt.BatchMapFn(ctx)
+	if err != nil {
+		go func() {
```

Comment:
We are trying to write to the channel, it will be a blocking call until we have a receiver end for the error channel.
We would be returning the error channel from this function itself. So as to not have a blocking situation in a case where the receiver has not started reading on the channel yet.